aws/acmpca.go

```diff
@@ -0,0 +1,120 @@
+package aws
+
+import (
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/acmpca"
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/hashicorp/go-multierror"
+)
+
+// getAllACMPCA returns a list of all arns of ACMPCA, which can be deleted.
+func getAllACMPCA(session *session.Session, region string, excludeAfter time.Time, configObj config.Config) ([]*string, error) {
+	svc := acmpca.New(session)
+	var arns []*string
+	if paginationErr := svc.ListCertificateAuthoritiesPages(&acmpca.ListCertificateAuthoritiesInput{}, func(p *acmpca.ListCertificateAuthoritiesOutput, lastPage bool) bool {
+		for _, ca := range p.CertificateAuthorities {
+			if shouldIncludeACMPCA(ca, excludeAfter, configObj) {
+				arns = append(arns, ca.Arn)
+			}
+		}
+		return !lastPage
+	}); paginationErr != nil {
+		return nil, errors.WithStackTrace(paginationErr)
+	}
+	return arns, nil
+}
+
+func shouldIncludeACMPCA(ca *acmpca.CertificateAuthority, excludeAfter time.Time, configObj config.Config) bool {
+	if ca == nil {
+		return false
+	}
+
+	// one can only delete CAs if they are 'ACTIVE' or 'DISABLED'
+	statusSafe := aws.StringValue(ca.Status)
+	isCandidateForDeletion := statusSafe == acmpca.CertificateAuthorityStatusActive || statusSafe == acmpca.CertificateAuthorityStatusDisabled
+	if !isCandidateForDeletion {
+		return false
+	}
+
+	// reference time for excludeAfter is lastStateChangeAt time,
+	// unless it was never changed and createAt time is used.
+	var referenceTime time.Time
+	if ca.LastStateChangeAt == nil {
+		referenceTime = aws.TimeValue(ca.CreatedAt)
+	} else {
+		referenceTime = aws.TimeValue(ca.LastStateChangeAt)
+	}
+	if excludeAfter.Before(referenceTime) {
+		return false
+	}
+
+	return config.ShouldInclude(
+		aws.StringValue(ca.Arn),
+		configObj.ACMPCA.IncludeRule.NamesRegExp,
+		configObj.ACMPCA.ExcludeRule.NamesRegExp,
```

Comment:
This is checking the names regex against the ARN, which is unintuitive. Is there a way to just get the name out?

Reply:
Yeah. I will think about it. You are correct that this does not feel good.
It should be something unique. And the `name` is not unique.
What is your plan about the `regex` behaviour for other resources? Is this "the final form"? Otherwise I would get rid of the regex in total.

Reply:
I would need your guidance on this. Indeed this would be the ARN and there is nothing better than using the ARN for it.

I could introduce new types?

```
type Config struct {
....
	ACMPCA                    ResourceTypeARN `yaml:"acmcpa"`
}

type ResourceTypeARN struct {
	IncludeRule FilterRuleARN `yaml:"include"`
	ExcludeRule FilterRuleARN `yaml:"exclude"`
}

type FilterRuleARN struct {
	ARNsRegExp []Expression `yaml:"arns_regex"`
}
```

Or I could do:

```
type FilterRule struct {
	NamesRegExp []Expression `yaml:"names_regex"`
	ARNsRegExp []Expression `yaml:"names_regex"`
}


```

Reply:
> What is your plan about the regex behaviour for other resources? Is this "the final form"? Otherwise I would get rid of the regex in total.

This is not the final form, as we plan on supporting tags (something users have requested for other resources). Adding in ARN support or tag support just for ACM PCA makes sense to me, although that would be a lot more work to add considering the potential regressions with other resources.

If it isn't critical for your environment, I might suggest removing the regex filters for now and add that in later as a future feature when we support something that makes more sense, like ARN or tags.

Reply:
Alright. I have removed the `configObj` support (see the latest commit)